src/main/scala/org/broadinstitute/hail/expr/Fun.scala

```diff
@@ -1,34 +1,114 @@
 package org.broadinstitute.hail.expr
 
 sealed trait Fun {
-  def retType: BaseType
+  def retType: Type
+
+  def subst(): Fun
 
   def convertArgs(transformations: Array[UnaryFun[Any, Any]]): Fun
 }
 
-case class UnaryFun[T, U](retType: BaseType, f: (T) => U) extends Fun with Serializable with ((T) => U) {
+case class Arity0Aggregator[T, U](retType: Type, ctor: () => TypedAggregator[U]) extends Fun {
+  def subst() = Arity0Aggregator[T, U](retType.subst(), ctor)
+
+  def convertArgs(transformations: Array[UnaryFun[Any, Any]]): Fun = {
+    assert(transformations.length == 1)
+
+    Arity0Aggregator[T, U](retType, () => new TransformedAggregator(ctor(), transformations(0).f))
+  }
+}
+
+class TransformedAggregator[T](val prev: TypedAggregator[T], transform: (Any) => Any) extends TypedAggregator[T] {
+  def seqOp(x: Any) = prev.seqOp(transform(x))
+
+  def combOp(agg2: this.type) = prev.combOp(agg2.prev)
+
+  def result = prev.result
+
+  def copy() = new TransformedAggregator(prev.copy(), transform)
+}
+
+case class Arity1Aggregator[T, U, V](retType: Type, ctor: (U) => TypedAggregator[V]) extends Fun {
+  def subst() = Arity1Aggregator[T, U, V](retType.subst(), ctor)
+
+  def convertArgs(transformations: Array[UnaryFun[Any, Any]]): Fun = {
+    assert(transformations.length == 2)
+
+    Arity1Aggregator[T, U, V](retType, { (u) =>
+      new TransformedAggregator(ctor(
+        transformations(1).f(u).asInstanceOf[U]),
```

Comment:
We have a registered conversion [from ints to doubles](https://github.com/hail-is/hail/blob/8493d1d668912e904f101fce85643ca4cd2b769b/src/main/scala/org/broadinstitute/hail/expr/FunctionRegistry.scala#L613). The `transformations` array erases the types to `Any => Any`, but we assume that `u` and `f(u)` have the same type, which isn't true in the case of a numeric conversion (unless `U == Number`).